lib/src/merged_tree.rs

```diff
@@ -692,24 +692,19 @@ impl Iterator for ConflictIterator {
 }
 
 /// Iterator over the differences between two trees.
+///
+/// Tree entries (`MergedTreeValue::is_tree()`) are included only if the other
+/// side is present and not a tree.
 pub struct TreeDiffIterator<'matcher> {
     store: Arc<Store>,
-    stack: Vec<TreeDiffItem>,
+    stack: Vec<TreeDiffDir>,
     matcher: &'matcher dyn Matcher,
 }
 
-struct TreeDiffDirItem {
+struct TreeDiffDir {
     entries: Vec<(RepoPathBuf, MergedTreeValue, MergedTreeValue)>,
 }
 
-enum TreeDiffItem {
-    Dir(TreeDiffDirItem),
-    // This is used for making sure that when a directory gets replaced by a file, we
-    // yield the value for the addition of the file after we yield the values
-    // for removing files in the directory.
-    File(RepoPathBuf, MergedTreeValue, MergedTreeValue),
-}
-
 impl<'matcher> TreeDiffIterator<'matcher> {
     /// Creates a iterator over the differences between two trees. Generally
     /// prefer `MergedTree::diff()` of calling this directly.
```

Comment:
nit: remove or update doc about `MergedTree::diff()` since `TreeDiffIterator::new()` is no longer identical to `MergedTree::diff_stream()`.